controllers/cloud_config_sync_controller_test.go

```diff
@@ -0,0 +1,275 @@
+				apierrors.IsNotFound(cl.Get(ctx, client.ObjectKeyFromObject(managedCloudConfig), &corev1.ConfigMap{})),
+			).Should(BeTrue())
+		}
+
+		if syncedCloudConfigMap != nil {
+			Expect(cl.Delete(ctx, syncedCloudConfigMap, deleteOptions)).To(Succeed())
+			Eventually(
+				apierrors.IsNotFound(cl.Get(ctx, client.ObjectKeyFromObject(syncedCloudConfigMap), &corev1.Namespace{})),
+			).Should(BeTrue())
+		}
+
+		infraCloudConfig = nil
+		managedCloudConfig = nil
+		syncedCloudConfigMap = nil
+
+		infra := makeInfrastrucutreResource()
+		Expect(cl.Delete(ctx, infra)).To(Succeed())
+		Eventually(
+			apierrors.IsNotFound(cl.Get(ctx, client.ObjectKeyFromObject(infra), infra)),
+		).Should(BeTrue())
+	})
+
+	It("config should be synced up after first reconcile", func() {
+		syncedCloudConfigMap := &corev1.ConfigMap{}
+		Eventually(func() error {
+			return cl.Get(ctx, syncedConfigMapKey, syncedCloudConfigMap)
+		}).Should(Succeed())
+		Expect(syncedCloudConfigMap.Data[defaultConfigKey]).Should(BeEquivalentTo("bar"))
+	})
+
+	It("config should be synced up if managed cloud config changed", func() {
+		changedManagedConfig := managedCloudConfig.DeepCopy()
+		changedManagedConfig.Data = map[string]string{"cloud.conf": "managed one changed"}
+		Expect(cl.Update(ctx, changedManagedConfig)).To(Succeed())
+
+		Eventually(func() (bool, error) {
+			syncedCloudConfigMap := &corev1.ConfigMap{}
+			err := cl.Get(ctx, syncedConfigMapKey, syncedCloudConfigMap)
+			if err != nil {
+				return false, err
+			}
+			return syncedCloudConfigMap.Data[defaultConfigKey] == "managed one changed", nil
+		}).Should(BeTrue())
+	})
+
+	It("config should be synced up if own cloud-config deleted or changed", func() {
+		syncedCloudConfigMap := &corev1.ConfigMap{}
+		Expect(cl.Get(ctx, syncedConfigMapKey, syncedCloudConfigMap)).Should(Succeed())
+
+		syncedCloudConfigMap.Data = map[string]string{"foo": "baz"}
+		Expect(cl.Update(ctx, syncedCloudConfigMap)).To(Succeed())
+		Eventually(func() (bool, error) {
+			err := cl.Get(ctx, syncedConfigMapKey, syncedCloudConfigMap)
+			if err != nil {
+				return false, err
+			}
+			return syncedCloudConfigMap.Data[defaultConfigKey] == "bar", nil
+		}).Should(BeTrue())
+
+		Expect(cl.Delete(ctx, syncedCloudConfigMap)).To(Succeed())
+		Eventually(func() (bool, error) {
+			err := cl.Get(ctx, syncedConfigMapKey, syncedCloudConfigMap)
+			if err != nil {
+				return false, err
+			}
+			return syncedCloudConfigMap.Data[defaultConfigKey] == "bar", nil
+		}).Should(BeTrue())
+	})
+
+	It("config should not be updated if source and target config content are identical", func() {
+		syncedCloudConfigMap := &corev1.ConfigMap{}
+		Expect(cl.Get(ctx, syncedConfigMapKey, syncedCloudConfigMap)).Should(Succeed())
+		initialCMresourceVersion := syncedCloudConfigMap.ResourceVersion
+
+		request := reconcile.Request{NamespacedName: client.ObjectKey{Name: "foo", Namespace: "bar"}}
+		_, err := reconciler.Reconcile(ctx, request)
+		Expect(err).Should(Succeed())
```

Comment:
The manager is handling this in the background, shouldn't need to call the reconciler directly ever

Reply:
Here I wanted to ensure that reconcile was definitely run and finished successfully and check things after. it's most convinient way which i found... Just for make this particular test scope bit more narrow.

Reply:
Bit of an odd pattern, maybe worth a comment to explain the reasoning as you've just said there